lib/std/sums.nim

```diff
@@ -53,13 +53,111 @@ func sumPairs*[T](x: openArray[T]): T =
   let n = len(x)
   if n == 0: T(0) else: sumPairwise(x, 0, n)
 
+func fastTwoSum*[T: SomeFloat](a, b: T): (T, T) =
+  ## Deker's algorithm
+  ## pre-condition: |a| >= |b|
+  ## you must swap a and b if pre-condition is not satisfied
+  ## s + r = a + b exactly (s is result[0] and r is result[1])
+  ## s is the nearest FP number of a + B
+  assert(abs(a) >= abs(b))
+  result[0] = a + b
+  result[1] = b - (result[0] - a)
+
+func twoSum*[T](a, b: T): (T, T) =
+  ## Møller-Knuth's algorithm.
+  ## Improve Deker's algorithm, no branch needed
+  ## More operations, but still cheap.
+  result[0] = a + b
+  let z = result[0] - a
+  result[1] = (a - (result[0] - z)) + (b - z)
+
+func sum2*[T: SomeFloat](v: openArray[T]): T =
+  ## sum an array v using twoSum function
+  if len(v) == 0:
+    return 0.0
+
+  var s = v[0]
+  var e: float
+  for i in 1..<v.len-1:
+    let sum = twoSum(s, v[i])
+    s = sum[0]
+    e += sum[1]
+  return s + e
+
+func sumShewchuck_add[T: SomeFloat](v: openArray[T]): seq[T] =
+  for x in v:
+    var x = x
+    var i = 0
+    for y in result:
+      let sum = twoSum(x, y)
+      let hi = sum[0]
+      let lo = sum[1]
+      if lo != 0.0:
+        result[i] = lo
+        i.inc
+      x = hi
+    setLen(result, i + 1)
+    result[i] = x
+
+func sumShewchuck_total[T: SomeFloat](partials: openArray[T]): T =
+  var hi = 0.0
+  if len(partials) > 0:
+    var n = len(partials)
+    dec(n)
+    hi = partials[n]
+    var lo = 0.0
+    while n > 0:
+      var x = hi
+      dec(n)
+      var y = partials[n]
+      let sum = twoSum(x, y)
+      hi = sum[0]
+      lo = sum[1]
+      if lo != 0.0:
+        break
+      if (n > 0 and
```

Comment:
if n > 0 and (lo < 0.0 and partials[n - 1] < 0.0) or
            (lo > 0.0 and partials[n - 1] > 0.0))